pkg/chartmuseum/server.go

```diff
@@ -207,94 +284,84 @@ func (server *Server) listObjectsGetDiff() ([]storage.Object, storage.ObjectSlic
 		}
 	}
 
-	diff := storage.GetObjectSliceDiff(server.StorageCache, filteredObjects)
-	return filteredObjects, diff, nil
+	return filteredObjects, nil
 }
 
-func (server *Server) regenerateRepositoryIndex() error {
-	server.Logger.Debugw("Acquiring storage cache lock")
-	server.StorageCacheLock.Lock()
-	server.Logger.Debugw("Storage cache lock acquired")
-	defer func() {
-		server.Logger.Debugw("Releasing storage cache lock")
-		server.StorageCacheLock.Unlock()
-	}()
-
-	objects, diff, err := server.listObjectsGetDiff()
-	if err != nil {
-		return err
-	}
-
-	index := &repo.Index{
-		IndexFile: server.RepositoryIndex.IndexFile,
-		Raw:       server.RepositoryIndex.Raw,
-		ChartURL:  server.RepositoryIndex.ChartURL,
-	}
+func (server *Server) regenerateRepositoryIndexWorker(diff storage.ObjectSliceDiff, storageObjects []storage.Object, reqID string) (*repo.Index, error) {
+	server.Logger.Debugf("(%s) Regenerating index.yaml", reqID)
+    index := &repo.Index{
+        IndexFile: server.RepositoryIndex.IndexFile,
+        Raw:       server.RepositoryIndex.Raw,
+        ChartURL:  server.RepositoryIndex.ChartURL,
+    }
 
 	for _, object := range diff.Removed {
-		err := server.removeIndexObject(index, object)
+		err := server.removeIndexObject(index, object, reqID)
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 
 	for _, object := range diff.Updated {
-		err := server.updateIndexObject(index, object)
+		err := server.updateIndexObject(index, object, reqID)
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	// Parallelize retrieval of added objects to improve startup speed
-	err = server.addIndexObjectsAsync(index, diff.Added)
+	// Parallelize retrieval of added objects to improve speed
+	err := server.addIndexObjectsAsync(index, diff.Added, reqID)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	server.Logger.Debug("Regenerating index.yaml")
 	err = index.Regenerate()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	// It is very important that these two stay in sync as they reflect the same reality. StorageCache serves
+	// as object modification time cache, and RepositoryIndex is the canonical cached index.
 	server.RepositoryIndex = index
-	server.StorageCache = objects
-	return nil
+	server.StorageCache = storageObjects
+
+	server.Logger.Debugf("(%s) index.yaml regenerated", reqID)
+	return index, nil
 }
 
-func (server *Server) removeIndexObject(index *repo.Index, object storage.Object) error {
+func (server *Server) removeIndexObject(index *repo.Index, object storage.Object, reqID string) error {
 	chartVersion, err := server.getObjectChartVersion(object, false)
 	if err != nil {
 		return server.checkInvalidChartPackageError(object, err, "removed")
 	}
-	server.Logger.Debugw("Removing chart from index",
+	server.Logger.Debugw(fmt.Sprintf("(%s) Removing chart from index", reqID),
```

Comment:
requestID should be passed the Debugw similar to name and version below (in all instances)

Reply:
Yes, this is how I implemented it first. But it made it harder to link the different requests together. Having a prefix request id makes it easy to see at a glance (without scanning to the right), the ids are all aligned.

The result is this:

```
(455) Removing chart from index ...
```

instead of this: 

```
Removing chart from index  { "Request ID": 455, "name" : "", "version": 0.0.0 }
```

I'll do what you prefer.

Reply:
Ok, let do both then. Its useful to be able to parse when using `--log-json`. Can we specify `"reqID": 455` vs `"Request ID": 455`?

Also we may want to send this back in a response header such as `X-ChartMuseum-ID` (we can do this later).

Reply:
ok, I'll do that.